### ISSUE TITLE
v0.4 switch changes needed

### ISSUE THREAD

**Issue by Arceliar** [yggdrasil-network/yggdrasil-go#658]
Sort-of a followup to #555.

Part way through v0.3, we added a set of queues to the switch. All packets are passed to the switch worker, which either forwards them immediately to an idle peer, or buffers them until a peer becomes idle. Initially, it would forward to *any* idle peer that is closer to the destination (in tree metric space), and buffer only if *all* of them are busy. However, this resulted in lots of packets arriving out-of-order, which higher layers of the network stack dislike (e.g. TCP throttles down to almost nothing).

Later on, we switched to sending to the "best" peer (minimum distance to the destination, with ties broken by minimum latency to the root) if that peer is idle, and otherwise buffering until a peer (any peer) becomes idle. This was done as a hotfix to make sure that (most) traffic tends to be forwarded to the "best" peer, since this prevents (most) packet reordering, which keeps the transport and application layers of the network stack much happier.

That hotfix needs to be removed in v0.4 and replaced with something better. There's basically 3 directions we could go:

1. Revert to forwarding across all paths, and then have some mechanism at the session level to reorder packets correctly (e.g. TCP-like retransmission or FEC). This adds a lot of complexity just to do the job that the higher layers of the stack are supposed to handle themselves, and even if we do this *perfectly* we'd end up killing performance for some applications (bandwidth would be great, but there'd be latency jitter every time a packet arrived out-of-order and needed to be buffered until the earlier packets arrive or are reconstructed from FEC). So this isn't really an option for now, as far as I'm concerned (it's too much work to implement something that isn't supposed to be our job in the first place, and it distracts from the part I actually care about -- compact routing).

2. Instead of a single set of queues in the switch, which all traffic goes through, keep a separate queue per peer. This would let us keep separate queues per traffic flow (so e.g. a bulk TCP transfer to one user doesn't block a small stream of UDP traffic from another user's VoIP call), so we'd get to keep the congestion control aspect of the current switch, we just wouldn't have the option to route around congested links (which we currently avoid anyway, because the higher layers of the stack hate it). The main benefit here is that it would simplify the switch logic enough that we can have each peer (actor) do switch lookups in parallel, so it removes a (theoretical) bottleneck from having a single central switch actor.

3. Exactly the same as 2, but we could take things a step further and use source routing. Long story short, ygg already suffers from almost all of the down-sides of source routing, and gains almost none of the benefits, so this is a very natural thing to do. It also simplifies a *lot* of other things I wanted to do in v0.4 (mostly related to DoS and DDoS mitigation).

I'm currently planning on doing 3, I just need to decide on a few details for the source routing. That means 2 will probably be implemented as an intermediate step first (to remove the switch worker from the hot loop), and then migrated to run on top of source routing a little later on.

**Comment by janicez** [comment]
does this break compatibility with previous Yggdrasils?

**Comment by Revertron** [comment]
Is it possible to just postpone read from TUN when all peers are congested?

P.S.: I like your idea number 2, but didn't understand "source routing" thing.

**Comment by neilalexander** [comment]
Source routing was added in v0.4, closing.
